File: packages/php/src/Webhook.php
<?php

declare(strict_types=1);

namespace Opensms;

/**
 * Webhook signature verification. Static, so a receiver can verify
 * deliveries without constructing a client or holding an API key.
 *
 * OpenSMS signs each delivery with the header
 * `X-OpenSMS-Signature: t=<unix seconds>,v1=<hex>`, where
 * `v1 = hex(HMAC-SHA256(secret, "<t>.<raw body>"))` and `secret` is the full
 * `whsec_...` value returned once by `webhooks->create()`. Parsing mirrors the
 * server (`internal/webhooks/signature.go`) exactly.
 */
final class Webhook
{
    public const SIGNATURE_HEADER = 'X-OpenSMS-Signature';
    public const DEFAULT_TOLERANCE = 300;

    private function __construct()
    {
    }

    /**
     * True when `$header` is a valid, unexpired signature of `$payload`.
     *
     * @param string $payload The exact raw request body bytes (verify before parsing JSON).
     * @param string $header The X-OpenSMS-Signature header value.
     * @param string $secret The endpoint secret, `whsec_...`, used verbatim.
     * @param array{toleranceSeconds?: int, now?: int} $options
     */
    public static function verifySignature(string $payload, string $header, string $secret, array $options = []): bool
    {
        return self::check($payload, $header, $secret, $options) === null;
    }

    /**
     * Verify the signature and decode the event envelope
     * `{id, type, workspace_id, environment, created_at, data}`.
     *
     * @param array{toleranceSeconds?: int, now?: int} $options
     * @return array<string, mixed>
     * @throws OpensmsException status 0 with code "invalid_signature" or "expired_signature"
     */
    public static function constructEvent(string $payload, string $header, string $secret, array $options = []): array
    {
        $failure = self::check($payload, $header, $secret, $options);
        if ($failure !== null) {
            $message = $failure === 'expired_signature'
                ? 'OpenSMS webhook signature timestamp is outside the tolerance.'
                : 'OpenSMS webhook signature is invalid.';

            throw new OpensmsException(0, $message, errorCode: $failure);
        }

        try {
            $event = json_decode($payload, true, 512, JSON_THROW_ON_ERROR);
        } catch (\JsonException $e) {
            throw new OpensmsException(0, 'OpenSMS webhook payload is not valid JSON.', errorCode: 'invalid_payload', previous: $e);
        }
        if (!is_array($event)) {
            throw new OpensmsException(0, 'OpenSMS webhook payload is not a JSON object.', errorCode: 'invalid_payload');
        }

        return $event;
    }

    /**
     * @param array{toleranceSeconds?: int, now?: int} $options
     * @return string|null null when valid, else "invalid_signature" or "expired_signature"
     */
    private static function check(string $payload, string $header, string $secret, array $options): ?string
    {
        $tolerance = (int) ($options['toleranceSeconds'] ?? self::DEFAULT_TOLERANCE);
        $now = (int) ($options['now'] ?? time());

        if (trim($secret) === '' || $tolerance < 0) {
            return 'invalid_signature';
        }

        $values = self::parseHeader($header);
        if ($values === null) {
            return 'invalid_signature';
        }

        $t = $values['t'];
        if (preg_match('/^[+-]?\d{1,19}$/', $t) !== 1) {
            return 'invalid_signature';
        }
        $timestamp = (int) $t;
        if (abs($now - $timestamp) > $tolerance) {
            return 'expired_signature';
        }

        $v1 = $values['v1'];
        if (strlen($v1) !== 64 || !ctype_xdigit($v1)) {
            return 'invalid_signature';
        }

        $expected = hash_hmac('sha256', $t . '.' . $payload, $secret);
        if (!hash_equals($expected, strtolower($v1))) {
            return 'invalid_signature';
        }

        return null;
    }

    /**
     * Split on ",", trim, split on the first "="; reject empty parts,
     * duplicates, and anything other than exactly `t` and `v1`.
     *
     * @return array{t: string, v1: string}|null
     */
    private static function parseHeader(string $header): ?array
    {
        $values = [];
        foreach (explode(',', $header) as $part) {
            $pair = explode('=', trim($part), 2);
            if (count($pair) !== 2 || $pair[0] === '' || $pair[1] === '' || isset($values[$pair[0]])) {
                return null;
            }
            $values[$pair[0]] = $pair[1];
        }
        if (count($values) !== 2 || !isset($values['t'], $values['v1'])) {
            return null;
        }

        return ['t' => $values['t'], 'v1' => $values['v1']];
    }
}

File: packages/rust/src/webhook.rs
//! Webhook signature verification.
//!
//! The API signs every delivery with
//! `X-OpenSMS-Signature: t=<unix seconds>,v1=<hex>`, where
//! `v1 = hex(HMAC-SHA256(secret, "<t>." + raw body))` and `secret` is the full
//! `whsec_...` string returned once by `webhooks.create`. Parsing mirrors the
//! server exactly. These functions need no client or API key.

use std::collections::HashMap;

use crate::crypto::{constant_time_eq, hmac_sha256, unhex};
use crate::error::{OpensmsError, Result};
use crate::models::WebhookEvent;
use crate::util::unix_now;

/// The header that carries the signature.
pub const SIGNATURE_HEADER: &str = "X-OpenSMS-Signature";

/// Default allowed clock skew, in seconds.
pub const DEFAULT_TOLERANCE_SECONDS: i64 = 300;

/// Options for [`verify_signature`] and [`construct_event`].
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct VerifyOptions {
    /// Maximum `|now - t|` in seconds (inclusive). Default 300.
    pub tolerance_seconds: i64,
    /// Current Unix time override, for tests. Default: the system clock.
    pub now: Option<i64>,
}

impl Default for VerifyOptions {
    fn default() -> Self {
        Self {
            tolerance_seconds: DEFAULT_TOLERANCE_SECONDS,
            now: None,
        }
    }
}

enum Failure {
    Invalid,
    Expired,
}

fn check(
    payload: &[u8],
    header: &str,
    secret: &str,
    opts: VerifyOptions,
) -> std::result::Result<(), Failure> {
    if secret.trim().is_empty() || opts.tolerance_seconds < 0 {
        return Err(Failure::Invalid);
    }
    let mut fields: HashMap<&str, &str> = HashMap::new();
    for part in header.split(',') {
        let part = part.trim();
        let (k, v) = part.split_once('=').ok_or(Failure::Invalid)?;
        if k.is_empty() || v.is_empty() || fields.insert(k, v).is_some() {
            return Err(Failure::Invalid);
        }
    }
    if fields.len() != 2 {
        return Err(Failure::Invalid);
    }
    let t_raw = *fields.get("t").ok_or(Failure::Invalid)?;
    let v1 = *fields.get("v1").ok_or(Failure::Invalid)?;
    let t: i64 = t_raw.parse().map_err(|_| Failure::Invalid)?;
    let now = opts.now.unwrap_or_else(unix_now);
    if (now - t).abs() > opts.tolerance_seconds {
        return Err(Failure::Expired);
    }
    if v1.len() != 64 {
        return Err(Failure::Invalid);
    }
    let given = unhex(v1).ok_or(Failure::Invalid)?;
    let mut message = Vec::with_capacity(t_raw.len() + 1 + payload.len());
    message.extend_from_slice(t_raw.as_bytes());
    message.push(b'.');
    message.extend_from_slice(payload);
    let expected = hmac_sha256(secret.as_bytes(), &message);
    if constant_time_eq(&expected, &given) {
        Ok(())
    } else {
        Err(Failure::Invalid)
    }
}

/// `true` when `header` is a valid, fresh signature of `payload` (the exact
/// raw body bytes) under `secret`.
pub fn verify_signature(payload: &[u8], header: &str, secret: &str, opts: VerifyOptions) -> bool {
    check(payload, header, secret, opts).is_ok()
}

/// Verify the signature, then parse the event envelope. Fails with an
/// [`OpensmsError`] whose `code` is `invalid_signature` or `expired_signature`.
pub fn construct_event(
    payload: &[u8],
    header: &str,
    secret: &str,
    opts: VerifyOptions,
) -> Result<WebhookEvent> {
    match check(payload, header, secret, opts) {
        Ok(()) => serde_json::from_slice(payload)
            .map_err(|e| OpensmsError::transport(format!("failed to decode webhook event: {e}"))),
        Err(Failure::Invalid) => Err(OpensmsError::signature(
            "invalid_signature",
            "webhook signature does not match",
        )),
        Err(Failure::Expired) => Err(OpensmsError::signature(
            "expired_signature",
            "webhook signature timestamp is outside the tolerance",
        )),
    }
}
